tests/extension/oneapi-current-device/current_device.cpp

```diff
@@ -0,0 +1,136 @@
+//  limitations under the License.
+//
+*******************************************************************************/
+
+#include "../../common/common.h"
+#include <condition_variable>
+#include <mutex>
+#include <thread>
+#include <vector>
+
+namespace current_device {
+#ifdef SYCL_EXT_ONEAPI_CURRENT_DEVICE
+
+sycl::device get_device_to_set(const bool is_need_to_reallocate) {
+  static std::vector<sycl::device> devices;  // = sycl::device::get_devices();
+  if (is_need_to_reallocate) {
+    devices = sycl::device::get_devices();
+  }
+  for (const auto device : devices) {
+    // remove device from the list of devices
+    devices.erase(std::remove(devices.begin(), devices.end(), device),
+                  devices.end());
+    return device;
+  }
+  SKIP("Test did not find any device to set as current");
+}
+
+#endif
+
+TEST_CASE(
+    "Test for \"get_current_device\" extension, get default device, call "
+    "function and compare devices ") {
+#ifndef SYCL_EXT_ONEAPI_CURRENT_DEVICE
+  SKIP("SYCL_EXT_ONEAPI_CURRENT_DEVICE is not defined");
+#else
+  if (sycl::device::get_devices().size() < 1) {
+    SKIP("Test requires at least one device");
+  }
+  const sycl::device default_device{sycl::default_selector()};
+  const auto current_device =
+      sycl::ext::oneapi::experimental::this_thread::get_current_device();
+  CHECK(
+      std::is_same_v<std::remove_cv_t<decltype(current_device)>, sycl::device>);
+  CHECK(default_device == current_device);
+#endif
+}
+
+TEST_CASE(
+    "Test for \"set_current_device and get_current_device\" extension, check "
+    "that get function returns the device set by set function") {
+#ifndef SYCL_EXT_ONEAPI_CURRENT_DEVICE
+  SKIP("SYCL_EXT_ONEAPI_CURRENT_DEVICE is not defined");
+#else
+  const auto device_to_set = current_device::get_device_to_set(true);
+  sycl::ext::oneapi::experimental::this_thread::set_current_device(
+      device_to_set);
+  const auto current_device =
+      sycl::ext::oneapi::experimental::this_thread::get_current_device();
+  CHECK(
+      std::is_same_v<std::remove_cv_t<decltype(current_device)>, sycl::device>);
+  CHECK(device_to_set == current_device);
+#endif
+}
+
+TEST_CASE(
+    "Test for \"set_current_device and get_current_device in different "
+    "threads\" extension, check "
+    "that each thread has its own current device") {
+#ifndef SYCL_EXT_ONEAPI_CURRENT_DEVICE
+  SKIP("SYCL_EXT_ONEAPI_CURRENT_DEVICE is not defined");
+#else
+  if (sycl::device::get_devices().size() < 2) {
+    SKIP("Test requires at least two devices");
+  }
+  std::mutex mtx;
+  std::condition_variable cv_thread1;
+  std::condition_variable cv_thread2;
+  bool signal_thread1 = false;
+  bool signal_thread2 = false;
+  const auto t1_device_to_set = get_device_to_set(true);
+  const auto t2_device_to_set = get_device_to_set(false);
+  sycl::device t1_current_device;
+  sycl::device t2_current_device;
+
+  std::thread t1([&]() {
+    {
+      std::unique_lock<std::mutex> lock(mtx);
```

Comment:
We should not use a mutex here. We need to make sure that we don't hide a potential thread safety problem in feature impl here, because by spec set_current_device and get_current_device must be thread safe.